Support RRef.backward() for Owner RRefs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46641 Support RRef.backward() for User RRefs.
* **#46568 Support RRef.backward() for Owner RRefs.**

This PR adds support for an RRef.backward() API. This would be useful
in applications like pipeline parallelism as described here:
https://github.com/pytorch/pytorch/issues/44827

This PR only adds support for owner RRefs, remote RRef support will be added in
a follow up PR.

Differential Revision: [D24406311](https://our.internmc.facebook.com/intern/diff/D24406311/)

pritamdamania committed:
Support RRef.backward() for local RRefs.

This PR adds support for an RRef.backward() API. This would be useful
in applications like pipeline parallelism as described here:
https://github.com/pytorch/pytorch/issues/44827

This PR only adds support for local RRefs, remote RRef support will be added in
a follow up PR.

Differential Revision: [D24406311](https://our.internmc.facebook.com/intern/diff/D24406311/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support RRef.backward() for user RRefs."


This PR adds support for an RRef.backward() API. This would be useful
in applications like pipeline parallelism as described here:
https://github.com/pytorch/pytorch/issues/44827

This PR only adds support for local RRefs, remote RRef support will be added in
a follow up PR.

Differential Revision: [D24406311](https://our.internmc.facebook.com/intern/diff/D24406311/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                  ctx_id starting from the owner of the RRef. In this case,
+                  :meth:`~torch.distributed.autograd.get_gradients` should be
+                  used to retrieve the gradients. If ``dist_autograd_ctx_id``
+                  is None, it is assumed that this is a local autograd graph
```
nit: double quote None

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                    dist_autograd_ctx_id (int, optional): The distributed
+                        autograd context id for which we should retrieve the
+                        gradients (default: -1).
+                    retain_graph(bool, optional): If False, the graph used to
```
nit: double quote `False`

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                        gradients (default: -1).
+                    retain_graph(bool, optional): If False, the graph used to
+                        compute the grad will be freed. Note that in nearly all
+                        cases setting this option to True is not needed and
```
ditto: double quote `True`

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                        compute the grad will be freed. Note that in nearly all
+                        cases setting this option to True is not needed and
+                        often can be worked around in a much more efficient way.
+                        Usually, you need to set this to True to run backward
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
 
+void PyRRef::backward(int64_t dist_autograd_ctx_id, bool retain_graph) {
+  if (rref_->isOwner()) {
+    auto value =
```
can this be `const&`?

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
+    TORCH_CHECK(
+        value.isTensor(), "RRef should contain a tensor for .backward()");
+    auto root = value.toTensor();
+    TORCH_CHECK(root.requires_grad(), "requires_grad not set on RRef's value");
+    TORCH_CHECK(
+        root.numel() == 1,
+        "RRef does not contain a scalar, all roots need to be scalar");
+    TORCH_CHECK(
+        root.grad_fn(), "RRef does not have a valid gradient function.");
```
Curious, do we need these three checks, or did the autograd engine already do that?

pritamdamania87 replied:
These were unnecessary, I called the CPP APIs instead and these checks were in that API.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3635,6 +3635,35 @@ def test_cannot_infer_backend_from_options(self):
+        rref.backward()
+        self.assertEqual(torch.ones_like(t), t.grad)
+
+        with dist_autograd.context() as context_id:
```
shall we add a test where part of the distributed autograd graph is on a different rank?

mrshenli reviewed (approved):
LGTM, added some minor comments.

rohan-varma commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
+          dist_autograd_ctx_id, {root}, retain_graph);
+    }
+  } else {
+    // TODO
```
Should we raise an error for now for UserRRefs or do we anticipate to land https://github.com/pytorch/pytorch/pull/46641#pullrequestreview-515149684 together/soon after this one?

pritamdamania87 replied:
I'd anticipate to land https://github.com/pytorch/pytorch/pull/46641 soon and don't think anyone would be using this in fbcode in the mean time.

pritamdamania committed:
Update on "Support RRef.backward() for Owner RRefs."


This PR adds support for an RRef.backward() API. This would be useful
in applications like pipeline parallelism as described here:
https://github.com/pytorch/pytorch/issues/44827

This PR only adds support for owner RRefs, remote RRef support will be added in
a follow up PR.

Differential Revision: [D24406311](https://our.internmc.facebook.com/intern/diff/D24406311/)

[ghstack-poisoned]
